### ISSUE TITLE
Adaptations to run script for experiments

### ISSUE THREAD

**Issue by jim-smith** [AI-SDC/SACRO-ML#14]
Based on @albacrespi 's script discussed [here](https://github.com/AI-SDC/SACRO-ML/discussions/11)

Changes needed (may well be more)
-[] Instead of saving 'random_state' to recreate train/test splits should save indices   
    to future-proof wrt changes in sklearn and also to allow subsampling
-[] The attack scripts from aisdc need to have option to save  the predicted probabilities from the attack model for train and test items.                 - maybe as an array of length (num_train+num_test) with  two columns: 0/1 for test/train, then predicted prob(train).
- I think this makes more sense than just saving the attack model (since e.g. LIRA doesn't have one) 
- and it also allows us to trace exactly which individuals were 'at risk' from a die nMI Attack

- [] should support adding/revising attacks later and then running them on the saved target models

**Comment by albacrespi** [comment]
I'm creating a checklist of things I'll be working on. My understanding so far is that some things will need to be added to Target.py, and others at each of the attack files separetly.

- [ ] Create a new parameter/index to control split. Might need to inntroduce it in all the attacks functions.
- [ ] Make sure all the times that "train_test_split" this parameter is used correctly.
- [ ] Create a function that if the target model is found, it loads it and recreates data split, otherwise create it all from scratch
- [ ] Add option to save attack probabilities.
- [ ] If attack probabilities are found load them and recalculate metrics.

Also I noticed an inconsistency in the way AIA and MIA attacks are spliting the data:
In the worst case is done like this :
    `train_test_split(X, y, test_size=0.3)`

in the AIA is done like this
    `train_test_split( x, y, test_size=0.5, stratify=y, shuffle=True)`

This should be consistent everywhere. Also, when making the split reproducible the variable random_state needs to be used. So I suggest to create a new function to be called by all the attacks which sets:
 - random_state = VALUE
 - stratify = Y
 - shuffle = TRUE
 - test_size = TEST_SIZE

Let me know if I'm missing something.

**Comment by albacrespi** [comment]
I'm currently adapting the code from GRAIMatter for the experiments.
When creating the target model, do we want to create several models with different splits of the data?
I noted that if this is true there would be N targets models x M attacks (on worst case scenario). 
I wonder if it's better not to have N target models, we might end up having a large number of complex data to interpret.
Any thoughts @jim-smith @simonrnss @shahzadmumtaz22 ?

**Comment by albacrespi** [comment]
![Experiments_design](https://github.com/AI-SDC/SACRO-ML/assets/15641866/908ff2c5-5fdc-4c10-b6ea-36b456348707)
This is a diagram of what I'm working on.

**Comment by jim-smith** [comment]
This looks good.
Re your comments about how we do the train_test_split, I definitely agree that we should stratify on Y.

**Comment by jim-smith** [comment]
@albacrespi 

So far, for the structural attacks I will be reporting these metrics:
- wb_DoF_risk: binary/boolean
- wb_k_anonymity_risk: binary or possibly integer 
- wb_class_disclosure_risk: binary/boolean
- bb_k_anonymity_risk: boolean
- bb_class_disclosure_risk
- unnecessary_risk: binary/boolean

where wb or bb refers to white or black box risk.
@albacrespi @simonrnss @rpreen  for some types of models (xgboost ...) we will not be able to do the white box values - not sure what we should report in those cases, None?

**Comment by albacrespi** [comment]
The **metrics.py** script has a function called "get_metrics". Anything in there gets automatically added to the attacks - at least for worst case and lira. In the cases where some metrics can't be reported, simply ignore them and they won't appear in the report or the experimental results.
